File: AllInOne/src/TwoPointers/SortColors.js
/**
 * 
 * 75. Sort Colors
 * 
 */
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    
  if (nums.length < 1 || nums.length > 300) return undefined;

  let p0 = 0, p2 = nums.length - 1, shift = 0;

  const swap = (left, right) => {
    let tmp = nums[left];
    nums[left] = nums[right];
    nums[right] = tmp;
  }

  while (shift <= p2) {
    if (nums[shift] === 0) {
      swap(shift ++, p0 ++);
    } else if (nums[shift] === 2) {
      swap(shift, p2 --);
    } else {
      shift ++;
    }
  }
};

File: AllInOne/src/TwoPointers/SortColors.java
package src.TwoPointers;

/**
 * 
 * 75. Sort Colors
 * 
 * @author jingjiejiang
 * @history May 6, 2021
 * 
 */
public class SortColors {
    public void sortColors(int[] nums) {
        
        assert nums != null && nums.length >= 1;

        int p0 = 0, p2 = nums.length - 1;
        int shift = 0;

        while (shift <= p2) {

            if (nums[shift] == 0) {
                swap(nums, shift ++, p0 ++);
            } else if (nums[shift] == 2) {
                swap(nums, shift, p2 --);
            } else {
                shift ++;
            }
        }
    }

    private void swap(int[] nums, int left, int right) {

        int temp = nums[left];
        nums[left] = nums[right];
        nums[right] = temp;
    }
}
